src/irc/core/irc-cap.c

```diff
@@ -0,0 +1,229 @@
+			server->cap_active = g_slist_remove(server->cap_active, caps[i]);
+			cap_emit_signal(server, "nak", caps[i]);
+		}
+	}
+
+	g_strfreev(caps);
+	g_free(params);
+}
+
+static void event_invalid_cap (IRC_SERVER_REC *server, const char *data, const char *from)
+{
+	/* The server didn't understand one (or more) requested caps, terminate the negotiation.
+	 * This could be handled in a graceful way but since it shouldn't really ever happen this seems a
+	 * good way to deal with 410 errors. */
+	server->cap_complete = FALSE;
+	irc_send_cmd_now(server, "CAP END");
+}
+
+static gboolean sasl_timeout (IRC_SERVER_REC *server)
+{
+	/* The authentication timeout, we can't do much beside terminating it */
+	g_critical("The authentication timed out, try increasing the timeout and check your connection "
+	           "to the network.");
+	irc_send_cmd_now(server, "AUTHENTICATE *");
+	cap_finish_negotiation(server);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void sasl_init (IRC_SERVER_REC *server, const char *data, const char *from)
+{
+	irc_send_cmd_now(server, "AUTHENTICATE PLAIN");
+	g_timeout_add(20 * 1000, (GSourceFunc) sasl_timeout, server);
+}
+
+static void sasl_fail (IRC_SERVER_REC *server, const char *data, const char *from)
+{
+	/* Stop any pending timeout, if any */
+	g_source_remove_by_funcs_user_data((GSourceFunc) sasl_timeout, NULL);
+
+	g_critical("Authentication failed, make sure your credentials are correct and that the mechanism "
+	           "you have selected is supported by this server.");
+
+	/* Terminate the negotiation */
+	cap_finish_negotiation(server);
+}
+
+static void sasl_already (IRC_SERVER_REC *server, const char *data, const char *from)
+{
+	/* We're already authenticate, do nothing */
+	cap_finish_negotiation(server);
+}
+
+static void sasl_success (IRC_SERVER_REC *server, const char *data, const char *from)
+{
+	/* The authentication succeeded, time to finish the CAP negotiation */
+	g_warning("SASL authentication succeeded");
+	cap_finish_negotiation(server);
+}
+
+static void sasl_step (IRC_SERVER_REC *server, const char *data, const char *from)
+{
+	IRC_SERVER_CONNECT_REC *conn;
+	GString *req;
+	char *enc_req;
+
+	conn = server->connrec;
+
+	/* Stop the timer */
+	g_source_remove_by_funcs_user_data((GSourceFunc) sasl_timeout, NULL);
+
+	/* At this point we assume that conn->{username, password} are non-NULL.
+	 * The PLAIN mechanism expects a NULL-separated string composed by the authorization identity, the
+	 * authentication identity and the password.
+	 * The authorization identity field is optional and can be omitted, the server will derive the
+	 * identity by looking at the credentials provided.
+	 * The whole request is then encoded in base64. */
+
+	req = g_string_new(NULL);
+
+	g_string_append_c(req, '\0');
+	g_string_append(req, conn->username);
+	g_string_append_c(req, '\0');
+	g_string_append(req, conn->password);
+
+	enc_req = g_base64_encode((const guchar *)req->str, req->len);
```

Comment:
Added in glib 2.12, current requirement is 2.6.

IMO this is a good reason to bump the requirement. @ahf?

RHEL 5, the oldest still-supported distro version i know of, has glib 2.12, coincidentally, so i imagine it would cost nothing to bump it